### PR TITLE
fix: if artifact GC Pod fails make sure error is propagated as a Condition

### PR DESCRIPTION
**Pull request by juliev0**
Partial fix to #9953 

Issue was that we were only adding the Condition based on failure of individual artifacts. In this case, the GC Pod never got to the point of attempting deletion of an artifact (it couldn't read what the artifacts were in fact), so there was no reporting of individual artifacts' status. Now a Condition will be written if the Pod fails as well.

Signed-off-by: Julie Vogelman <julie_vogelman@intuit.com>

Fixes #TODO

Please do not open a pull request until you have checked ALL of these:

* [ ] Create the PR as draft .
* [ ] Run `make pre-commit -B` to fix codegen and lint problems.
* [ ] Sign-off your commits (otherwise the DCO check will fail).
* [ ] Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).
* [ ] "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
* [ ] Add unit or e2e tests. Say how you tested your changes. If you changed the UI, attach screenshots.
* [ ] Github checks are green.
* [ ] Once required tests have passed, mark your PR "Ready for review".

If changes were requested, and you've made them, dismiss the review to get it reviewed again.

### CODE REVIEW

File: workflow/controller/artifact_gc.go [hunk 1]
```diff
@@ -541,42 +541,57 @@ func (woc *wfOperationCtx) allArtifactsDeleted() bool {
 func (woc *wfOperationCtx) processCompletedArtifactGCPod(ctx context.Context, pod *corev1.Pod) error {
 	woc.log.Infof("processing completed Artifact GC Pod '%s'", pod.Name)
 
+	strategyStr, found := pod.Annotations[common.AnnotationKeyArtifactGCStrategy]
+	if !found {
+		return fmt.Errorf("Artifact GC Pod '%s' doesn't have annotation '%s'?", pod.Name, common.AnnotationKeyArtifactGCStrategy)
+	}
+	strategy := wfv1.ArtifactGCStrategy(strategyStr)
```
Comment (juliev0) [inline workflow/controller/artifact_gc.go:548]: note this code is just moved up from below

File: workflow/controller/artifact_gc.go [hunk 2]
```diff
@@ -541,42 +541,57 @@ func (woc *wfOperationCtx) allArtifactsDeleted() bool {
...
 		if err != nil {
 			return err
 		}
+		if allArtifactsSucceeded && pod.Status.Phase == corev1.PodSucceeded {
+			// now we can delete it, if it succeeded (otherwise we leave it up to be inspected)
+			woc.log.Debugf("deleting WorkflowArtifactGCTask: %s", task.Name)
+			err := woc.controller.wfclientset.ArgoprojV1alpha1().WorkflowArtifactGCTasks(woc.wf.Namespace).Delete(ctx, task.Name, metav1.DeleteOptions{})
+			if err != nil {
+				woc.log.Errorf("error deleting WorkflowArtifactGCTask: %s: %v", task.Name, err)
+			}
```
Comment (juliev0) [inline workflow/controller/artifact_gc.go:574]: this deletion code (which deletes both the WorkflowArtifactGCTask, as well as the Pod which it owns, used to just be in `processCompletedWorkflowArtifactGCTask()` but really this logic is dependent on both the Pod succeeding and the artifacts succeeding

File: workflow/controller/artifact_gc.go [hunk 3]
```diff
@@ -541,42 +541,57 @@ func (woc *wfOperationCtx) allArtifactsDeleted() bool {
...
+func (woc *wfOperationCtx) processCompletedWorkflowArtifactGCTask(artifactGCTask *wfv1.WorkflowArtifactGCTask, strategy wfv1.ArtifactGCStrategy) (bool, error) {
 	woc.log.Debugf("processing WorkflowArtifactGCTask %s", artifactGCTask.Name)
 
 	foundGCFailure := false
 	for nodeName, nodeResult := range artifactGCTask.Status.ArtifactResultsByNode {
 		// find this node result in the Workflow Status
 		wfNode, found := woc.wf.Status.Nodes[nodeName]
 		if !found {
-			return fmt.Errorf("node named '%s' returned by WorkflowArtifactGCTask '%s' wasn't found in Workflow '%s' Status", nodeName, artifactGCTask.Name, woc.wf.Name)
+			return false, fmt.Errorf("node named '%s' returned by WorkflowArtifactGCTask '%s' wasn't found in Workflow '%s' Status", nodeName, artifactGCTask.Name, woc.wf.Name)
```
Comment (terrytangyuan) [inline workflow/controller/artifact_gc.go:591]: Use %q?

Comment (juliev0) [reply workflow/controller/artifact_gc.go:591]: had to look that up. I guess it gives you a quoted string - good idea

Comment (juliev0) [reply workflow/controller/artifact_gc.go]: okay, just did a find and replace for all instances of `'%s'`

**Review by terrytangyuan — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**